### PR TITLE
[Disk Manager]: Snapshot: Check chunk contents for zeroes

### PR DESCRIPTION
**Pull request by jkuradobery**
Sometimes NBS can return chunks full of zeroes yet the zero field is set to false. Check if chunk is actually a zero chunk.

### CODE REVIEW

File: cloud/disk_manager/internal/pkg/dataplane/common/chunk_test.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package common
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/common/chunk_test.go:8]: ////////////////////////////////////////////////////////////////////////////////

File: cloud/disk_manager/internal/pkg/dataplane/common/chunk.go [hunk 2]
```diff
@@ -1,11 +1,14 @@
 package common
 
 import (
+	"bytes"
 	"hash/crc32"
 )
 
 ////////////////////////////////////////////////////////////////////////////////
 
+var zeroChunk = make([]byte, 1024*1024)
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/common/chunk.go:10]: zeroBytes

Comment (SvartMetal) [reply cloud/disk_manager/internal/pkg/dataplane/common/chunk.go:10]: or zeroData or zeroes

File: cloud/disk_manager/internal/pkg/dataplane/common/chunk.go [hunk 3]
```diff
@@ -18,3 +21,17 @@ type Chunk struct {
 func (chunk Chunk) Checksum() uint32 {
 	return crc32.ChecksumIEEE(chunk.Data)
 }
+
+func (chunk Chunk) IsZero() bool {
+	for i := 0; i < len(chunk.Data); i += len(zeroChunk) {
+		endOffset := i + len(zeroChunk)
+		if endOffset > len(chunk.Data) {
+			endOffset = len(chunk.Data)
+		}
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/common/chunk.go:30]: \n

File: cloud/disk_manager/internal/pkg/dataplane/common/chunk.go [hunk 4]
```diff
@@ -18,3 +21,17 @@ type Chunk struct {
 func (chunk Chunk) Checksum() uint32 {
 	return crc32.ChecksumIEEE(chunk.Data)
 }
+
+func (chunk Chunk) IsZero() bool {
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/common/chunk.go:25]: CheckDataIsAllZeroes

File: cloud/disk_manager/internal/pkg/dataplane/snapshot/storage/storage_ydb_impl.go [hunk 5]
```diff
@@ -590,7 +590,7 @@ func (s *storageYDB) WriteChunk(
 	useS3 bool,
 ) (chunkID string, err error) {
 
-	if chunk.Zero {
+	if chunk.Zero || chunk.IsZero() {
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/snapshot/storage/storage_ydb_impl.go:593]: it should be checked when reading chunk

File: cloud/disk_manager/internal/pkg/dataplane/common/chunk.go [hunk 6]
```diff
@@ -18,3 +21,17 @@ type Chunk struct {
...
+	for i := 0; i < len(chunk.Data); i += len(zeroChunk) {
+		endOffset := i + len(zeroChunk)
+		if endOffset > len(chunk.Data) {
+			endOffset = len(chunk.Data)
+		}
+		chunkPart := chunk.Data[i:endOffset]
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/common/chunk.go:33]: dataPart

File: cloud/disk_manager/internal/pkg/dataplane/common/chunk_test.go [hunk 7]
```diff
@@ -12,11 +12,11 @@ func TestChunkIsZero(t *testing.T) {
 	chunk := Chunk{
 		Data: []byte{1, 2, 3, 4, 5},
 	}
-	require.False(t, chunk.IsZero())
+	require.False(t, chunk.CheckDataIsAllZeroes())
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/common/chunk_test.go:15]: test should have the same name

TestCheckChunkDataIsAllZeroes

File: cloud/disk_manager/internal/pkg/dataplane/snapshot/storage/storage_ydb_test.go [hunk 8]
```diff
@@ -1262,6 +1262,40 @@ func TestCompressionMetricsCollection(t *testing.T) {
 	}
 }
 
+func TestSnapshotReadWriteAllZeroedChunk(t *testing.T) {
```
Comment (SvartMetal) [inline cloud/disk_manager/internal/pkg/dataplane/snapshot/storage/storage_ydb_test.go:1265]: still needed?

**Review by SvartMetal — APPROVED** [review]

**Comment by jkuradobery** [comment]
Closing since I made a new PR from a non-main branch in my repo. This PR waits for long-running tests, keeping a PR in my repo's main blocks my developement.

**Pull request closed without merging**